experimental/internal/taxa/names.go

```diff
@@ -0,0 +1,236 @@
+
+package taxa
+
+var (
+	// names is an array of user-visible names of all of the productions in this
+	// package.
+	names = [...]string{
+		Unknown:      "<unknown>",
+		Unrecognized: "unrecognized token",
+		TopLevel:     "file scope",
+		EOF:          "end-of-file",
+
+		Decl:         "declaration",
+		Empty:        "empty declaration",
+		Syntax:       "`syntax` declaration",
+		Edition:      "`edition` declaration",
+		Package:      "`package` declaration",
+		Import:       "import",
+		WeakImport:   "weak import",
+		PublicImport: "public import",
+		Extensions:   "extension range",
+		Reserved:     "reserved range",
+		Body:         "definition body",
+
+		Def:     "definition",
+		Message: "message definition",
+		Enum:    "enum definition",
+		Service: "service definition",
+		Extend:  "message extension block",
+		Oneof:   "oneof definition",
+
+		Option:       "option setting",
+		CustomOption: "custom option setting",
+
+		Field:     "message field",
+		EnumValue: "enum value",
+		Method:    "service method",
+
+		CompactOptions: "compact options",
+		MethodIns:      "method parameter list",
+		MethodOuts:     "method return type",
+
+		FieldTag:    "message field tag",
+		OptionValue: "option setting value",
+
+		Path:            "path",
+		ExtensionInPath: "path in `(...)`",
+
+		Expr:      "expression",
+		Range:     "range expression",
+		Array:     "array expression",
+		Dict:      "dictionary expression",
+		DictField: "key-value pair",
+
+		Type:       "type",
+		TypeParams: "type parameters",
+
+		Whitespace: "whitespace",
+		Comment:    "comment",
+		Ident:      "identifier",
+		String:     "string literal",
+		Float:      "floating-point literal",
+		Int:        "integer literal",
+
+		Semicolon: "`;`",
+		Comma:     "`,`",
+		Slash:     "`/`",
+		Colon:     "`:`",
+		Equals:    "`=`",
+		Minus:     "`-`",
+		Period:    "`.`",
+
+		LParen:   "`(`",
+		LBracket: "`[`",
+		LBrace:   "`{`",
+		LAngle:   "`<`",
+
+		RParen:   "`)`",
+		RBracket: "`]`",
+		RBrace:   "`}`",
+		RAngle:   "`>`",
+
+		Parens:   "`(...)`",
+		Brackets: "`[...]`",
+		Braces:   "`{...}`",
+		Angles:   "`<...>`",
+
+		KeywordSyntax:  "`syntax`",
+		KeywordEdition: "`edition`",
+		KeywordImport:  "`import`",
+		KeywordWeak:    "`weak`",
+		KeywordPublic:  "`public`",
+		KeywordPackage: "`package`",
+
+		KeywordOption:  "`option`",
+		KeywordMessage: "`message`",
+		KeywordEnum:    "`enum`",
+		KeywordService: "`service`",
+		KeywordExtend:  "`extend`",
+		KeywordOneof:   "`oneof`",
+
+		KeywordExtensions: "`extensions`",
+		KeywordReserved:   "`reserved`",
+		KeywordTo:         "`to`",
+		KeywordRPC:        "`rpc`",
+		KeywordReturns:    "`returns`",
+
+		KeywordOptional: "`optional`",
+		KeywordRepeated: "`repeated`",
+		KeywordRequired: "`required`",
+		KeywordGroup:    "`group`",
+		KeywordStream:   "`stream`",
+	}
+
+	// constNames is an array of the names of the constants, fo ruse in GoString().
+	constNames = [...]string{
```

Comment:
This seems kind of awful to maintain. We definitely need a test to iterate over everything in `names` to make sure we've got a value in here. Otherwise it seems like we could trivially add a new name and forget to add it here and end up in trouble later if we somehow don't exercise the new subject in a test.

Reply:
There is already a test for this. Unfortunately it is *very* annoying that we can't generate this automatically. (Well, we could but I don't think writing a generate script is at all worth it...)